### PR TITLE
[#5566] Ensure using a ChannelInitializer via ServerBootstrap.handler…

### PR DESCRIPTION
**Pull request by normanmaurer**
…(...) produce correct ordering.

Motivation:

When a ChannelInitializer is used via ServerBootstrap.handler(...) the users handlers may be added after the internal ServerBootstrapAcceptor. This should not happen.

Modifications:

Delay the adding of the ServerBootstrapAcceptor until the initChannel(....) method returns.

Result:

Correct order of handlers in the ServerChannels ChannelPipeline.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @trustin @Spikhalskiy FyI

File: transport/src/main/java/io/netty/bootstrap/ServerBootstrap.java [hunk 1]
```diff
@@ -169,13 +169,23 @@ void init(Channel channel) throws Exception {
...
                 if (handler != null) {
                     pipeline.addLast(handler);
                 }
-                pipeline.addLast(new ServerBootstrapAcceptor(
-                        currentChildGroup, currentChildHandler, currentChildOptions, currentChildAttrs));
+
+                // We add this handler via the EventLoop as the user may have used a ChannelInitializer as handler.
+                // In this case the initChannel(...) method will only be called after this method returns. Because
+                // of this we need to ensure we add our handler in a delayed fashion so all the users handler are
+                // places in front of the ServerBootstrapAcceptor.
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/bootstrap/ServerBootstrap.java]: `places` -> `placed` (or inserted)

File: transport/src/test/java/io/netty/bootstrap/ServerBootstrapTest.java [hunk 2]
```diff
@@ -59,4 +66,74 @@ public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
...
+        testParentHandler(false);
+    }
+
+    @Test(timeout = 3000)
+    public void testParentHandlerViaChannelInitializer() throws Exception {
+        testParentHandler(true);
+    }
+
+    private static void testParentHandler(boolean channelInitializer) throws Exception {
+        final LocalAddress addr = new LocalAddress("test");
```
Comment (Scottmitch) [inline transport/src/test/java/io/netty/bootstrap/ServerBootstrapTest.java]: should we just use `InetSocketAddress(0)` so there is less chance of collision with other tests?

Comment (normanmaurer) [inline transport/src/test/java/io/netty/bootstrap/ServerBootstrapTest.java]: Its a `LocalAddress` not an `InetSocketAddress(...)`.

Comment (Scottmitch) [inline transport/src/test/java/io/netty/bootstrap/ServerBootstrapTest.java]: if we run tests in parallel, and another test uses this same address (`LocalAddress("test")`) is there a chance the tests will conflict with each other?

Comment (johnou) [inline transport/src/test/java/io/netty/bootstrap/ServerBootstrapTest.java]: yes 

Exception in thread "main" io.netty.channel.ChannelException: address already in use by: [id: 0xccd33fb2, L:local:test]

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (4638df2) and 4.0 (091551d)

File: transport/src/test/java/io/netty/bootstrap/ServerBootstrapTest.java (hunk 2 above)
Comment (Scottmitch) [inline transport/src/test/java/io/netty/bootstrap/ServerBootstrapTest.java]: thanks @johnou for verifying

**Pull request closed without merging**